File: lintcode_algo_2021/s7.-ds-design/130.Heapify/130_heapify__shifup.py
class Solution:
    """
    @param: A: Given an integer array
    @return: nothing
    """
    def heapify(self, A):
        # write your code here
        for i in range(len(A)):
            self.shift_up(A, i)
    
    def shift_up(self, A, i):
        while i != 0:
            father = (i - 1) // 2
            if A[i] > A[father]:
                break
            A[i], A[father] = A[father], A[i]
            i = father
        

File: lintcode_algo_2021/s7.-ds-design/130.Heapify/130_heapify__shifup.java
public class Solution {
    /*
     * @param A: Given an integer array
     * @return: nothing
     */
    public void heapify(int[] A) {
        // write your code here
        for (int i = 0; i < A.length; i++) {
            shiftUp(A, i);
        }
    }
    private void shiftUp(int[] A, int i) {
        while (i != 0) {
            int father = (i - 1)/2;
            if (A[i] > A[father]) {
                break;
            }
            int tmp = A[i];
            A[i] = A[father];
            A[father] = tmp;

            i = father;
        }
    }
}